rohmu/atomic_open.py

```diff
@@ -0,0 +1,40 @@
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Any, BinaryIO, cast, ContextManager, Generator, IO, Optional, TextIO
+
+import os
+
+
+def atomic_open_for_writing_text(final_path: Path, encoding: Optional[str] = None) -> ContextManager[TextIO]:
+    return cast(ContextManager[TextIO], atomic_opener(final_path, encoding=encoding, mode="w"))
+
+
+def atomic_open_for_writing_bytes(final_path: Path) -> ContextManager[BinaryIO]:
+    return cast(ContextManager[BinaryIO], atomic_opener(final_path, mode="wb"))
+
+
+@contextmanager
+def atomic_opener(
+    final_path: Path, mode: str, encoding: Optional[str] = None, _inode_spy=lambda inode: None
+) -> Generator[IO[Any], None, None]:
+    """
+    Creates a file object for writing which will only appear on the filesystem if the context manager succeeds.
+
+    It's designed to prevent partial writes (file will either be available with full content, or will be unavailable)
+    and the need for manual cleanup.
+    """
+    parent_dir = final_path.parent
+    if not parent_dir.exists():
+        raise IOError(f"Parent directory '{parent_dir}' must exist but is missing")
+    if "w" not in mode:
+        raise ValueError("Write mode must be used to make actual sense")
+
+    fd = os.open(str(parent_dir), os.O_TMPFILE | os.O_RDWR, 0o600)
+    file_obj = os.fdopen(fd, mode=mode, encoding=encoding)
+    try:
+        yield file_obj
+        path = "/proc/self/fd/{0}".format(fd)
+        _inode_spy(os.stat(path).st_ino)
```

Comment:
You can get the `fd` from the `file_obj` by doing  `file_obj.fileno()` and then you can call `os.stat` of that fd directly without using the path:
 - No need to construct and know about the linux `/proc/self/...` path 
 - No need for a test-specific hook since you already can reach the `fd` from the `file_obj`

Reply:
This one I'm not understanding. I need the ```/proc/self/...``` to create the hardlink to the actual filesystem, this is straight from manpage:

```
 snprintf(path, PATH_MAX,  "/proc/self/fd/%d", fd);
                  linkat(AT_FDCWD, path, AT_FDCWD, "/path/for/file",
                                          AT_SYMLINK_FOLLOW);
```

Also, I need the test-specific hook because I want to make sure the inode from the created file is actually the same of the file which gets created inside the block. This is to verify that a copy doesn't happen, since it's the same inode.

I'm not sure how you'd use the fd number to perform the same check. File descriptors are process-local and there's no guarantee of non-reuse AFAIK and, if they were, I'd risk testing implementation details rather than the interface.

Reply:
I forgot about the `path` in the link :clown_face: So yes, the path has to stay.

However, for the test, what I meant is in your test you can do:
```python
    with atomic_open_for_writing_bytes(final_path) as tmp_file:
        tmp_inode = os.stat(tmp_file.fileno()).st_ino
   assert tmp_inode == os.stat(final_path).st_ino
```
You're still comparing the inode, not comparing the fd.